### DISCUSSION TITLE
Defining PRG_TEST task to run my unit tests

### DISCUSSION THREAD

**Discussion by lpillsbury** [tcunit/TcUnit#173] [Q&A] [answered]
Hi everyone, I just wrote my first PLC program and now I want it to have unit tests (I know, I know I should have done it the other way around-- but I didn't know about TcUnit when I started). Anyways, I'm ok with the TcUnit syntax and setup--except how do I run my tests? I already have a MAIN, and when I Activate Configuration/Build that's what runs. I've followed the tcunit.org documentation and @sagatowski's videos pretty carefully (thank you for this) and my best guess is the line in the intro user's guide that says "we need to have a program PRG_TEST defined in a task that we can run locally on our engineering PC". 

I see only MAIN in my PlcTask in the Solution Explorer and I don't know how to get PRG_TEST there also (if that's what I need). Below is a screen shot of my Solution Explorer--all the tests are grayed out for probably the same reason. What am I missing? Thanks in advance

![image](https://user-images.githubusercontent.com/26933372/149025078-8538d106-c85e-45e9-b186-6e16ac64c547.png)

**Comment by sagatowski** [comment, marked as answer]
Hi Leah!

The PRG_TEST is just a program just like any other where you instantiate all your test suites. It doesn't have to be named PRG_TEST, you can name it whatever you want (it was just a suggestion). For the tests to execute however, the PRG_TEST (or whatever program you have your tests declared in) need to be linked to a task.

There are two ways to go about this:
1. Split the project into a library project + executable project. You can either create a separate library project that only includes all your function blocks and all your tests, and then create a library out of this that is used in the main executable project. The library will then only have one program for the unit tests (PRG_TEST or whatever name you prefer).

2. Have everything in one project (as above). Then you need to create a separate task (TEST-Task for example) just for the tests. Link this task to the PRG_TEST. If you don't want your tests to execute everytime you run your main project (which is recommended) you should make sure to disable the task prior to deploying the actual executable software.

I hope I was clear with my description. If you have any further questions, please ask.

ps.
You seem to be doing really cool stuff!
ds.

**Reply by lpillsbury** [reply]
Hi Jakob,
Thanks for the explanation. I decided to go for option 2 since I didn't know how to make a library, and for now the project doesn't have more parts. I think I understood that you meant I should make a new task like this:
![image](https://user-images.githubusercontent.com/26933372/149212761-fc55da86-c054-43ff-b827-5e11dac3c40f.png)

But then I couldn't figure out how to link PRG_TEST to it. What I ended up doing instead was to link PRG_TEST to the PLC task in addition to MAIN, and it worked for the purpose of running the tests.
![image](https://user-images.githubusercontent.com/26933372/149213146-2d24514d-e336-42f4-802d-4eea06152e52.png)

 I admit it's not elegant since I'm simultaneously running Main and the tests, and I'd be interested to know what the procedure would be to have a completely separate task for Test only, and how to enable and disable tasks individually before runtime.
 Thanks again,
 Leah

**Reply by sagatowski** [reply]
Hi @lpillsbury! I'll write a proper description of exactly what you want to achieve before this week ends.

Also regarding creating libraries I'm currently in the process of doing a YouTube tutorial/course of how to do this.
I have a TwinCAT 3 course/tutorial on YouTube (100% free) here: https://www.youtube.com/playlist?list=PLimaF0nZKYHz3I3kFP4myaAYjmYk1SowO

The next part (part 11) will be about library creation. I should be done with this part in 2-3 weeks.

I'll update this discussion once I have the F.A.Q. updated with all information you need.

**Reply by lpillsbury** [reply]
Awesome! looking forward to it.

**Comment by sagatowski** [comment]
I just realized that this is not 100% clear in the current documentation. I will add an entry to the [F.A.Q. of TcUnit](https://tcunit.org/frequently-asked-questions/) before the end of the week.
